crud/common/rebalance.lua

```diff
@@ -0,0 +1,197 @@
+        if trig == safe_mode_bucket_trigger then
+            box.space._bucket:on_replace(nil, trig)
+        end
+    end
+    rebalance.safe_mode = true
+
+    -- This function is running inside on_commit trigger, need pcall to protect from errors in external code.
+    pcall(rebalance.on_safe_mode_toggle.run, rebalance.on_safe_mode_toggle, true)
+
+    log.info('Rebalance safe mode enabled')
+end
+
+local function _safe_mode_disable()
+    -- We have disabled safe mode so we need to add the trigger to detect the beginning
+    -- of rebalance process to enable safe mode again.
+    box.space._bucket:on_replace(safe_mode_bucket_trigger)
+    rebalance.safe_mode = false
+
+    -- This function is running inside on_commit trigger, need pcall to protect from errors in external code.
+    pcall(rebalance.on_safe_mode_toggle.run, rebalance.on_safe_mode_toggle, false)
+
+    log.info('Rebalance safe mode disabled')
+end
+
+local function create_space()
+    local safe_mode_space = box.schema.space.create(SAFE_MODE_SPACE, {
+        engine = 'memtx',
+        format = {
+            { name = 'key', type = 'string' },
+            { name = 'value', type = 'any' },
+        },
+        is_local = true,
+        if_not_exists = true,
+    })
+    safe_mode_space:create_index('primary', { parts = { 'key' }, if_not_exists = true })
+    safe_mode_space:insert{ 'status', false }
+end
+
+local function create_trigger()
+    box.space[SAFE_MODE_SPACE]:on_replace(function()
+        box.on_commit(function(rows_iter)
+            local safe_space_id = box.space[SAFE_MODE_SPACE].id
+            -- There may be multiple operations on safe mode status tuple in one transaction.
+            -- We will take only the last action.
+            -- 0 = do nothing, 1 = enable safe mode, -1 = disable safe mode
+            local safe_mode_action = 0
+            for _, old, new, sp in rows_iter() do
+                if sp ~= safe_space_id then
+                    goto continue
+                end
+                assert((old == nil or old.key == 'status') and (new.key == 'status'))
+
+                if (not old or not old.value) and new.value then
+                    safe_mode_action = 1
+                elseif old.value and not new.value then
+                    safe_mode_action = -1
+                end
+
+                ::continue::
+            end
+
+            if safe_mode_action == 1 then
+                _safe_mode_enable()
+            elseif safe_mode_action == -1 then
+                _safe_mode_disable()
+            end
+        end)
+    end)
+end
+
+function rebalance.init()
+    local stored_safe_mode
+    if not box.info.ro then
+        if box.space[SAFE_MODE_SPACE] == nil then
+            create_space()
+            create_trigger()
+        end
+        stored_safe_mode = box.space[SAFE_MODE_SPACE]:get{ 'status' }
+    else
+        while box.space[SAFE_MODE_SPACE] == nil or box.space[SAFE_MODE_SPACE].index[0] == nil do
```

Comment:
What if none of the instances are in `RW`, when `rebalance.init` is called. All of them will just indefinitely wait for replication of the space. We need to wait either for rw or for space here

Reply:
Thanks! Added check that node became `RW`.